# Evaluation of short-read sequencing performance across the Mtb genome

This repository contains the analysis pipeline and coded needed to reproduce the analysis for [*Genomic sequence characteristics and the empiric accuracy of short-read sequencing, 2021, BioRxiv*](https://www.biorxiv.org/content/10.1101/2021.04.08.438862v1). All key bioinformatics processing is implemented using the [SnakeMake](https://snakemake.github.io/) workflow language with software dependencies defined for all steps using [Conda](https://docs.conda.io/en/latest/). Further data analysis is implemented in Jupyter notebooks in the DataAnalysis/ directory.

This repository is distributed under the [MIT license terms](LICENSE).




## Summary 

In this work we used long read sequencing to throughly benchmark Illumina sequencing performance across the Mtb genome. From this work we produced many useful insights for future studies of the Mtb genome using Illumina WGS.







# Installation

To install the dependances needed to reproduce analysis 







## Overview of dataset used

We used a dataset of 36 clinical Mtb isolates from lineages 1-6, and 8 which all had completed circularized genome assemblies.
Each isolate had both Illumina and PacBio WGS available. 







## Bioinformatics Data Processing (Using SnakeMake workflow engine)
















