# Evaluation of short-read sequencing performance across the Mtb genome

This repository contains the analysis pipeline and coded needed to reproduce the analysis for [*Genomic sequence characteristics and the empiric accuracy of short-read sequencing, 2021, BioRxiv*](https://www.biorxiv.org/content/10.1101/2021.04.08.438862v1). All key bioinformatics processing is implemented using the [SnakeMake](https://snakemake.github.io/) workflow language, along with downstream analysis avaialble in Jupyter notebooks using Python 3.7. All software dependencies defined for all processing and analysis steps using [Conda](https://docs.conda.io/en/latest/). 


## Contents
- [Key Results](#Installation)
- [Installation](#Installation)


## Key Results 

In this work we used long read sequencing to throughly benchmark Illumina sequencing performance across the Mtb genome. From this work we produced many useful insights for future studies of the Mtb genome using Illumina WGS.


## Useful Genome-wide statistics and visualizations (H37Rv, the Mtb reference genome)
From this work we present many useful results that can help guide future genomics studies of the Mtb genome using Illumina WGS. 



## Installation

All dependancesis needed to reproduce analysis can be installed via the Conda(https://docs.conda.io/en/latest/).
```
```
This environemnt will include the Snakemake workflow engine as well as key Python libraries for downstream analysis.



## A) Bioinformatics Pipeline (Assemby, alignment, and QC of Illumina and PacBio data)


## B) Supporting Data Analysis 



## Overview of dataset used

We used a dataset of 36 clinical Mtb isolates from lineages 1-6, and 8 which all had completed circularized genome assemblies.




## License
This repository is distributed under the [MIT license terms](LICENSE).

