<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.7.1 -->
<title>Evaluation of short-read sequencing performance across the Mtb genome | [“Evaluation of short-read sequencing performance across the Mtb genome”]</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Evaluation of short-read sequencing performance across the Mtb genome" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="[“Genomic sequence characteristics and the empiric accuracy of short-read sequencing (Marin et al. 2021)”]" />
<meta property="og:description" content="[“Genomic sequence characteristics and the empiric accuracy of short-read sequencing (Marin et al. 2021)”]" />
<link rel="canonical" href="http://localhost:4000/" />
<meta property="og:url" content="http://localhost:4000/" />
<meta property="og:site_name" content="[“Evaluation of short-read sequencing performance across the Mtb genome”]" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Evaluation of short-read sequencing performance across the Mtb genome" />
<script type="application/ld+json">
{"description":"[“Genomic sequence characteristics and the empiric accuracy of short-read sequencing (Marin et al. 2021)”]","headline":"Evaluation of short-read sequencing performance across the Mtb genome","url":"http://localhost:4000/","@type":"WebSite","name":"[“Evaluation of short-read sequencing performance across the Mtb genome”]","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <link rel="stylesheet" href="/assets/css/style.css?v=15bcef5bd6cf3254bafb8d00c71bbcb97a504a77">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Evaluation of short-read sequencing performance across the Mtb genome</h1>
      <h2 class="project-tagline">Genomic sequence characteristics and the empiric accuracy of short-read sequencing (Marin et al. 2021)</h2>
      
        <a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation" class="btn">View on GitHub</a>
      
      
    </section>

    <section class="main-content">
      <h1 id="evaluation-of-short-read-sequencing-performance-across-the-mtb-genome">Evaluation of short-read sequencing performance across the Mtb genome</h1>

<p>This repository contains the bioinformatics pipeline and code needed to reproduce the analysis for <a href="https://www.biorxiv.org/content/10.1101/2021.04.08.438862v1"><em>Genomic sequence characteristics and the empiric accuracy of short-read sequencing, 2021, BioRxiv</em></a>. All key bioinformatics processing is implemented using the <a href="https://snakemake.github.io/">SnakeMake</a> workflow system, along with all downstream analysis available in Jupyter notebooks using Python 3.7. All software dependencies are defined for processing and analysis steps using <a href="https://docs.conda.io/en/latest/">Conda</a>.</p>

<h2 id="contents">Contents</h2>
<ul>
  <li><a href="#Installation">Installation</a></li>
  <li><a href="#Processing-of-Illumina-and-PacBio-sequencing">SnakeMake Pipeline</a></li>
  <li><a href="#Supporting-Data-Analysis">Data Analysis</a></li>
  <li><a href="#Results">Results</a></li>
  <li><a href="#License">License</a></li>
</ul>

<h1 id="installation">Installation</h1>
<p>All dependencies needed to reproduce this analysis can be installed via <a href="https://docs.conda.io/en/latest/">Conda</a> .</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># 1) Clone repository
git clone https://github.com/farhat-lab/mtb-illumina-wgs-evaluation

# 2) Create a conda environment named 'CoreEnv_MtbEval_V1'
cd mtb-illumina-wgs-evaluation/

conda env create --file CondaEnvs/Core_CondaEnv_Mtb_Illumina_WGS_Benchmarking_V1.yml -n CoreEnv_MtbEval_V1

# 3) Activate environment (used for SnakeMake pipeline and data analysis)
conda activate CoreEnv_MtbEval_V1
</code></pre></div></div>

<p>The above command will create a Conda environment with:</p>
<ul>
  <li>a) the Snakemake workflow engine,</li>
  <li>b) bioinformatics software dependencies,</li>
  <li>c) Python libraries for downstream analysis.</li>
</ul>

<h1 id="processing-of-illumina-and-pacbio-sequencing">Processing of Illumina and PacBio sequencing</h1>

<p>The provided Snakemake pipeline implements all steps related to Assembly, alignment, and other key processing steps.</p>

<p>To run the SnakeMake pipeline you need to provide:</p>
<ul>
  <li>a) A TSV with sample metadata and input FASTQ paths (Illumina and PacBio). Examples can be found in <code class="language-plaintext highlighter-rouge">Data/201202_PMP_SM_50CI_AllDataSets_InputSeqDataPaths</code>.</li>
  <li>b) Path to a target outout directory</li>
  <li>c) A <a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation/blob/main/Snakemake_Rules_And_Config/config_PMP_V6.json">config file</a> with paths to H37Rv reference files (<a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_000962.3">NC_000962.3</a>) in FASTA and GBK formats.</li>
  <li>d) A <a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation/blob/main/Snakemake_Rules_And_Config/clusterConfig_PMP_V10.json">cluster config file</a> specifing computational resource requirements for steps of the pipeline when using <a href="https://slurm.schedmd.com/documentation.html">SLURM</a></li>
</ul>

<p>To run the SnakeMake workflow, run the following bash commands:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Define path to the target output directory
targetOutput_Dir="../Mtb_PacBio_And_Illumina_Manuscript_SnakeMake_Output_V1"

# Define config file 
SnakeMake_ConfigFile="Snakemake_Rules_And_Config/config_PMP_V6.json"

# Define SLURM config file
SLURM_Cluster_Config="Snakemake_Rules_And_Config/clusterConfig_PMP_V10.json"

# Define path to TSV that specifies Sample Metadata and FASTQ PATHs

inputData_TSV_Dir="Data/201202_PMP_SM_50CI_AllDataSets_InputSeqDataPaths"
input_SampleInfo_TSV="${inputData_TSV_Dir}/201202_MTb_50CI_Peru_ChinerOms_Ngabonziza_TBPortals_PacBioDatasetsMerged_SampleInfo_InputFQs.tsv"

# Run SnakeMake pipeline
mkdir ${targetOutput_Dir}
mkdir -p ${targetOutput_Dir}/O2logs/cluster/

snakemake -s SnakeFile_Main_Processing.smk.py --config output_dir=${targetOutput_Dir} inputSampleData_TSV=${input_SampleInfo_TSV} --configfile ${inputConfigFile} -p --use-conda -j 50 --cluster-config  ${SLURM_Cluster_Config}  --cluster "sbatch -p {cluster.p} -n {cluster.n}  -t {cluster.t} --mem {cluster.mem} -o ${targetOutput_Dir}/{cluster.o} -e ${targetOutput_Dir}/{cluster.e}" --latency-wait 35 -k 
</code></pre></div></div>
<p>The final command will begin the SnakeMake pipeline using the SLURM workload manager.</p>

<h1 id="supporting-data-analysis">Supporting Data Analysis</h1>

<p>The <a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation/tree/main/DataAnalysis">DataAnalysis/</a> directory contains Jupyter notebooks for downstream data processing, table generation, and figure generation related to this work.</p>

<h1 id="results">Results</h1>

<h3 id="useful-genome-wide-statistics-and-visualizations-across-h37rv-the-mtb-reference-genome">Useful Genome-wide statistics and visualizations (across <a href="(https://www.ncbi.nlm.nih.gov/nuccore/NC_000962.3)">H37Rv</a>, the Mtb reference genome)</h3>
<p>From this work we present many useful results that can help guide future genomics studies of the Mtb genome using Illumina WGS.</p>

<div>
    <iframe src="./jbrowse2/index.html" style="border: 1px solid black" width="100%" height="500">
    </iframe>
</div>

<h3 id="genome-masking-schemes"><a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation/tree/main/References/Mtb_H37Rv_MaskingSchemes">Genome Masking Schemes</a></h3>
<p>1) Refined Low Confidence (RLC) Regions (<a href="https://raw.githubusercontent.com/farhat-lab/mtb-illumina-wgs-evaluation/main/References/Mtb_H37Rv_MaskingSchemes/RLC_Regions.H37Rv.bed">RLC_Regions.H37Rv.bed</a>)</p>

<p>2) Low Pileup Mappability Regions (K=50bp, E=4 mismatches, <a href="https://raw.githubusercontent.com/farhat-lab/mtb-illumina-wgs-evaluation/main/References/Mtb_H37Rv_MaskingSchemes/201027_PMap_K50E4_Regions_BELOW_1.bed">201027_PMap_K50E4_Regions_BELOW_1.bed</a>)</p>

<p>3) RLC Regions &amp; Low Pileup Mappability Regions <strong>combined</strong> (<a href="https://raw.githubusercontent.com/farhat-lab/mtb-illumina-wgs-evaluation/main/Results/B_Extra_UsefulDataFiles/F_Defining_RLC_Regions/RLC_Regions.Plus.LowPmapK50E4.H37Rv.bed">RLC_Regions.Plus.LowPmapK50E4.H37Rv.bed</a>)</p>


      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/farhat-lab/mtb-illumina-wgs-evaluation">mtb-illumina-wgs-evaluation</a> is maintained by <a href="https://github.com/farhat-lab">farhat-lab</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </section>

    
  </body>
</html>
