(this["webpackJsonp@jbrowse/web"]=this["webpackJsonp@jbrowse/web"]||[]).push([[30],{2381:function(e,t,r){"use strict";var n=r(2382);e.exports=n},2382:function(e,t,r){"use strict";var n=r(92),i=n(r(429)),a=n(r(166)),o=n(r(192)),s=n(r(2383)),c=function(){function e(){(0,a.default)(this,e)}return(0,o.default)(e,[{key:"toString",value:function(){var e="left"===this.MateDirection?"]":"[";return"left"===this.Join?"".concat(e).concat(this.MatePosition).concat(e).concat(this.Replacement):"".concat(this.Replacement).concat(e).concat(this.MatePosition).concat(e)}}]),e}(),l=function(){function e(t){var r=this;if((0,a.default)(this,e),!t||!t.header||!t.header.length)throw new Error("empty header received");var n=t.header.split(/[\r\n]+/).filter((function(e){return e}));if(!n.length)throw new Error("no non-empty header lines specified");if(this.Breakend=c,this.strict=void 0===t.strict||t.strict,this.metadata=JSON.parse(JSON.stringify({INFO:s.default.InfoFields,FORMAT:s.default.GenotypeFields,ALT:s.default.AltTypes,FILTER:s.default.FilterTypes})),n.forEach((function(e){if(!e.startsWith("#"))throw new Error("Bad line in header:\n".concat(e));if(e.startsWith("##"))r._parseMetadata(e);else if(e){var t=e.split("\t"),n=t.slice(0,8),i=["#CHROM","POS","ID","REF","ALT","QUAL","FILTER","INFO"];if(t.length<8)throw new Error("VCF header missing columns:\n".concat(e));if(9===t.length)throw new Error("VCF header has FORMAT but no samples:\n".concat(e));if(n.length!==i.length||!n.every((function(e,t){return e===i[t]})))throw new Error("VCF column headers not correct:\n".concat(e));r.samples=t.slice(9)}})),!this.samples)throw new Error("VCF does not have a header line")}return(0,o.default)(e,[{key:"_parseMetadata",value:function(e){var t=e.trim().match(/^##(.+?)=(.*)/).slice(1,3),r=(0,i.default)(t,2),n=r[0],a=r[1];if(a.startsWith("<")){n in this.metadata||(this.metadata[n]={});var o=this._parseStructuredMetaVal(a),s=(0,i.default)(o,2),c=s[0],l=s[1];this.metadata[n][c]=l}else this.metadata[n]=a}},{key:"_parseStructuredMetaVal",value:function(e){var t=this._parseKeyValue(e.replace(/^<|>$/g,""),","),r=t.ID;return delete t.ID,"Number"in t&&(Number.isNaN(Number(t.Number))||(t.Number=Number(t.Number))),[r,t]}},{key:"getMetadata",value:function(){for(var e=this.metadata,t=arguments.length,r=new Array(t),n=0;n<t;n++)r[n]=arguments[n];for(var i=0;i<r.length;i+=1)if(!(e=e[r[i]]))return e;return e}},{key:"_parseKeyValue",value:function(e){for(var t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:";",r={},n="",i="",a=1,o=0;o<e.length;o+=1)1===a?"="===e[o]?a=2:e[o]!==t?n+=e[o]:""===i&&(r[n]=null,n=""):2===a?e[o]===t?(r[n]=i,n="",i="",a=1):'"'===e[o]?a=3:i+=e[o]:3===a&&('"'!==e[o]?i+=e[o]:a=2);return 2===a||3===a?r[n]=i:1===a&&(r[n]=null),r}},{key:"parseLine",value:function(e){var t=this;if((e=e.trim()).length){for(var r=0,n=0;r<e.length&&("\t"===e[r]&&(n+=1),9!==n);r+=1);var i=e.substr(0,r).split("\t"),a=e.substr(r+1),o={CHROM:i[0],POS:Number(i[1]),ID:"."===i[2]?null:i[2].split(";"),REF:i[3],ALT:"."===i[4]?null:i[4].split(","),QUAL:"."===i[5]?null:parseFloat(i[5])};if("."===i[6]?o.FILTER=null:"PASS"===i[6]?o.FILTER="PASS":o.FILTER=i[6].split(";"),this.strict&&void 0===i[7])throw new Error("no INFO field specified, must contain at least a '.' (turn off strict mode to allow)");var s=void 0===i[7]||"."===i[7]?{}:this._parseKeyValue(i[7]);Object.keys(s).forEach((function(e){var r;r=s[e]?(r=s[e].split(",")).map((function(e){return"."===e?null:e})):s[e];var n=t.getMetadata("INFO",e,"Type");n&&("Integer"===n||"Float"===n?r=r.map((function(e){return null===e?null:Number(e)})):"Flag"===n&&(s[e]?console.warn("Info field ".concat(e," is a Flag and should not have a value (got value ").concat(s[e],")")):r=!0)),s[e]=r})),o.INFO=s,o.ALT&&s&&s.SVTYPE&&"BND"===s.SVTYPE[0]&&(o.ALT=o.ALT.map(this._parseBreakend.bind(this)));var c=this;return Object.defineProperty(l.prototype,"SAMPLES",{get:function(){var e=c._parseGenotypes(i[8],a);return Object.defineProperty(this,"SAMPLES",{value:e}),e}}),new l(o)}function l(e){Object.assign(this,e)}}},{key:"_parseBreakend",value:function(e){var t=e.split(/[[\]]/);if(t.length>1){var r=new c;r.MateDirection=e.includes("[")?"right":"left";for(var n=0;n<t.length;n+=1){var i=t[n];i&&(i.includes(":")?(r.MatePosition=i,r.Join=r.Replacement?"right":"left"):r.Replacement=i)}return r}return e}},{key:"_parseGenotypes",value:function(e,t){var r=this;t=t.split("\t");var n={};return e=e&&e.split(":"),this.samples.forEach((function(i,a){n[i]={},e.forEach((function(e){n[i][e]=null})),t[a].split(":").forEach((function(t,a){var o;if(""===t||"."===t||void 0===t)o=null;else{o=t.split(",").map((function(e){return"."===e?null:e}));var s=r.getMetadata("FORMAT",e[a],"Type");"Integer"!==s&&"Float"!==s||!o||(o=o.map((function(e){return e?Number(e):null})))}n[i][e[a]]=o}),{})})),n}}]),e}();e.exports=l},2383:function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;t.default={InfoFields:{AA:{Number:1,Type:"String",Description:"Ancestral allele"},AC:{Number:"A",Type:"Integer",Description:"Allele count in genotypes, for each ALT allele, in the same order as listed"},AD:{Number:"R",Type:"Integer",Description:"Total read depth for each allele"},ADF:{Number:"R",Type:"Integer",Description:"Read depth for each allele on the forward strand"},ADR:{Number:"R",Type:"Integer",Description:"Read depth for each allele on the reverse strand"},AF:{Number:"A",Type:"Float",Description:"Allele frequency for each ALT allele in the same order as listed (estimated from primary data, not called genotypes)"},AN:{Number:1,Type:"Integer",Description:"Total number of alleles in called genotypes"},BQ:{Number:1,Type:"Float",Description:"RMS base quality"},CIGAR:{Number:1,Type:"Float",Description:"Cigar string describing how to align an alternate allele to the reference allele"},DB:{Number:0,Type:"Flag",Description:"dbSNP membership"},DP:{Number:1,Type:"Integer",Description:"combined depth across samples"},END:{Number:1,Type:"Integer",Description:"End position (for use with symbolic alleles)"},H2:{Number:0,Type:"Flag",Description:"HapMap2 membership"},H3:{Number:0,Type:"Flag",Description:"HapMap3 membership"},MQ:{Number:1,Type:null,Description:"RMS mapping quality"},MQ0:{Number:1,Type:"Integer",Description:"Number of MAPQ == 0 reads"},NS:{Number:1,Type:"Integer",Description:"Number of samples with data"},SB:{Number:4,Type:"Integer",Description:"Strand bias"},SOMATIC:{Number:0,Type:"Flag",Description:"Somatic mutation (for cancer genomics)"},VALIDATED:{Number:0,Type:"Flag",Description:"Validated by follow-up experiment"},"1000G":{Number:0,Type:"Flag",Description:"1000 Genomes membership"},IMPRECISE:{Number:0,Type:"Flag",Description:"Imprecise structural variation"},NOVEL:{Number:0,Type:"Flag",Description:"Indicates a novel structural variation"},SVTYPE:{Number:1,Type:"String",Description:"Type of structural variant"},SVLEN:{Number:null,Type:"Integer",Description:"Difference in length between REF and ALT alleles"},CIPOS:{Number:2,Type:"Integer",Description:"Confidence interval around POS for imprecise variants"},CIEND:{Number:2,Type:"Integer",Description:"Confidence interval around END for imprecise variants"},HOMLEN:{Type:"Integer",Description:"Length of base pair identical micro-homology at event breakpoints"},HOMSEQ:{Type:"String",Description:"Sequence of base pair identical micro-homology at event breakpoints"},BKPTID:{Type:"String",Description:"ID of the assembled alternate allele in the assembly file"},MEINFO:{Number:4,Type:"String",Description:"Mobile element info of the form NAME,START,END,POLARITY"},METRANS:{Number:4,Type:"String",Description:"Mobile element transduction info of the form CHR,START,END,POLARITY"},DGVID:{Number:1,Type:"String",Description:"ID of this element in Database of Genomic Variation"},DBVARID:{Number:1,Type:"String",Description:"ID of this element in DBVAR"},DBRIPID:{Number:1,Type:"String",Description:"ID of this element in DBRIP"},MATEID:{Number:null,Type:"String",Description:"ID of mate breakends"},PARID:{Number:1,Type:"String",Description:"ID of partner breakend"},EVENT:{Number:1,Type:"String",Description:"ID of event associated to breakend"},CILEN:{Number:2,Type:"Integer",Description:"Confidence interval around the inserted material between breakend"},DPADJ:{Type:"Integer",Description:"Read Depth of adjacency"},CN:{Number:1,Type:"Integer",Description:"Copy number of segment containing breakend"},CNADJ:{Number:null,Type:"Integer",Description:"Copy number of adjacency"},CICN:{Number:2,Type:"Integer",Description:"Confidence interval around copy number for the segment"},CICNADJ:{Number:null,Type:"Integer",Description:"Confidence interval around copy number for the adjacency"}},GenotypeFields:{AD:{Number:"R",Type:"Integer",Description:"Read depth for each allele"},ADF:{Number:"R",Type:"Integer",Description:"Read depth for each allele on the forward strand"},ADR:{Number:"R",Type:"Integer",Description:"Read depth for each allele on the reverse strand"},DP:{Number:1,Type:"Integer",Description:"Read depth"},EC:{Number:"A",Type:"Integer",Description:"Expected alternate allele counts"},FT:{Number:1,Type:"String",Description:'Filter indicating if this genotype was "called"'},GL:{Number:"G",Type:"Float",Description:"Genotype likelihoods"},GP:{Number:"G",Type:"Float",Description:"Genotype posterior probabilities"},GQ:{Number:1,Type:"Integer",Description:"Conditional genotype quality"},GT:{Number:1,Type:"String",Description:"Genotype"},HQ:{Number:2,Type:"Integer",Description:"Haplotype quality"},MQ:{Number:1,Type:"Integer",Description:"RMS mapping quality"},PL:{Number:"G",Type:"Integer",Description:"Phred-scaled genotype likelihoods rounded to the closest integer"},PQ:{Number:1,Type:"Integer",Description:"Phasing quality"},PS:{Number:1,Type:"Integer",Description:"Phase set"}},AltTypes:{DEL:{Description:"Deletion relative to the reference"},INS:{Description:"Insertion of novel sequence relative to the reference"},DUP:{Description:"Region of elevated copy number relative to the reference"},INV:{Description:"Inversion of reference sequence"},CNV:{Description:"Copy number variable region (may be both deletion and duplication)"},"DUP:TANDEM":{Description:"Tandem duplication"},"DEL:ME":{Description:"Deletion of mobile element relative to the reference"},"INS:ME":{Description:"Insertion of a mobile element relative to the reference"},NON_REF:{Description:"Represents any possible alternative allele at this location"},"*":{Description:"Represents any possible alternative allele at this location"}},FilterTypes:{PASS:{Description:"Passed all filters"}}}},2516:function(e,t,r){"use strict";r.r(t),r.d(t,"default",(function(){return N}));var n=r(88),i=r(87),a=r.n(i),o=r(91),s=r(86),c=r(90),l=r(199),u=r(164),p=r(93),f=r(95),d=r(127),h=r(129),m=r(140),b=r(816),g=r(2381),y=r.n(g),v=r(81),D=r(750),N=function(e){Object(p.a)(r,e);var t=Object(f.a)(r);function r(){var e;Object(s.a)(this,r);for(var n=arguments.length,i=new Array(n),a=0;a<n;a++)i[a]=arguments[a];return(e=t.call.apply(t,[this].concat(i))).configured=void 0,e}return Object(c.a)(r,[{key:"configure",value:function(){var e=Object(o.a)(a.a.mark((function e(){var t,r,n,i,o,s,c=this;return a.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return this.configured||(t=Object(v.readConfObject)(this.config,"vcfGzLocation"),r=Object(v.readConfObject)(this.config,["index","location"]),n=Object(v.readConfObject)(this.config,["index","indexType"]),i=Object(h.openLocation)(t),o="CSI"===n,s=new b.TabixIndexedFile({filehandle:i,csiFilehandle:o?Object(h.openLocation)(r):void 0,tbiFilehandle:o?void 0:Object(h.openLocation)(r),chunkCacheSize:50*Math.pow(2,20),chunkSizeLimit:1e9}),this.configured=s.getHeader().then((function(e){return{filehandle:i,vcf:s,parser:new y.a({header:e})}})).catch((function(e){throw c.configured=void 0,e}))),e.abrupt("return",this.configured);case 2:case"end":return e.stop()}}),e,this)})));return function(){return e.apply(this,arguments)}}()},{key:"getRefNames",value:function(){var e=Object(o.a)(a.a.mark((function e(){var t,r,n,i=arguments;return a.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return t=i.length>0&&void 0!==i[0]?i[0]:{},e.next=3,this.configure();case 3:return r=e.sent,n=r.vcf,e.abrupt("return",n.getReferenceSequenceNames(t));case 6:case"end":return e.stop()}}),e,this)})));return function(){return e.apply(this,arguments)}}()},{key:"getHeader",value:function(){var e=Object(o.a)(a.a.mark((function e(){var t,r;return a.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,this.configure();case 2:return t=e.sent,r=t.vcf,e.abrupt("return",r.getHeader());case 5:case"end":return e.stop()}}),e,this)})));return function(){return e.apply(this,arguments)}}()},{key:"getMetadata",value:function(){var e=Object(o.a)(a.a.mark((function e(){var t,r;return a.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,this.configure();case 2:return t=e.sent,r=t.parser,e.abrupt("return",r.getMetadata());case 5:case"end":return e.stop()}}),e,this)})));return function(){return e.apply(this,arguments)}}()},{key:"getFeatures",value:function(e){var t=this,r=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{};return Object(m.ObservableCreate)(function(){var i=Object(o.a)(a.a.mark((function i(o){var s,c,l,u,p,f;return a.a.wrap((function(i){for(;;)switch(i.prev=i.next){case 0:return s=e.refName,c=e.start,l=e.end,i.next=3,t.configure();case 3:return u=i.sent,p=u.vcf,f=u.parser,i.next=8,p.getLines(s,c,l,Object(n.a)({lineCallback:function(e,r){o.next(new D.a({variant:f.parseLine(e),parser:f,id:"".concat(t.id,"-vcf-").concat(r)}))}},r));case 8:o.complete();case 9:case"end":return i.stop()}}),i)})));return function(e){return i.apply(this,arguments)}}(),r.signal)}},{key:"getFeaturesInMultipleRegions",value:function(e){var t=this,n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},i=Object(l.a)(Object(u.a)(r.prototype),"getFeaturesInMultipleRegions",this);return Object(m.ObservableCreate)(function(){var r=Object(o.a)(a.a.mark((function r(o){var s,c,l,u,p;return a.a.wrap((function(r){for(;;)switch(r.prev=r.next){case 0:return r.next=2,t.bytesForRegions(e);case 2:return s=r.sent,r.next=5,t.configure();case 5:return c=r.sent,l=c.filehandle,r.next=9,l.stat();case 9:u=r.sent,(p=Math.round(s/u.size*100))>100&&(p=100),p>60&&console.warn("getFeaturesInMultipleRegions fetching ".concat(p,"% of VCF file, but whole-file streaming not yet implemented")),i.call(t,e,n).subscribe(o);case 14:case"end":return r.stop()}}),r)})));return function(e){return r.apply(this,arguments)}}())}},{key:"bytesForRegions",value:function(){var e=Object(o.a)(a.a.mark((function e(t){var r,n,i,o;return a.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,this.configure();case 2:return r=e.sent,n=r.vcf,e.next=6,Promise.all(t.map((function(e){return n.index.blocksForRange(e.refName,e.start,e.end)})));case 6:return i=e.sent,o=[],i.forEach((function(e){e.forEach((function(e){var t=e.minv.blockPosition,r=e.maxv.blockPosition+64e3;o.find((function(e){return e.start<=r&&e.end>=t&&(e.start=Math.min(e.start,t),e.end=Math.max(e.end,r),!0)}))||o.push({start:t,end:r})}))})),e.abrupt("return",o.reduce((function(e,t){return e+t.end-t.start+1}),0));case 10:case"end":return e.stop()}}),e,this)})));return function(t){return e.apply(this,arguments)}}()},{key:"freeResources",value:function(){}}]),r}(d.BaseFeatureDataAdapter)}}]);
//# sourceMappingURL=30.6b55744f.chunk.js.map